frontend/src/router.ts

```diff
@@ -180,46 +181,84 @@ export const router = new Router({
   ]
 });
 
+export const hasClaim = (
+  requiredClaims: { [key: string]: any },
+  userClaims: { [key: string]: any }
+) => {
+  return Object.entries(requiredClaims).some(
+    ([claimName, claimValue]) => userClaims[claimName] === claimValue
+  );
+};
+
+export const filterRoutesByClaims = (
+  routes: RouteConfig[] = [],
+  userClaims: { [key: string]: any },
+  requireParentClaims: boolean | { [key: string]: any } = false
+) => {
+  return routes.reduce(
+    (filteredRoutes, route) => {
+      const requireClaims = _.get(route, "meta.auth.requireClaims", false);
+
+      if (route.meta.activator) {
+        const childRoutes = filterRoutesByClaims(
+          route.children,
+          userClaims,
+          requireClaims || requireParentClaims
+        );
+        childRoutes.length &&
+          filteredRoutes.push({ ...route, children: childRoutes });
+      } else {
+        if (requireClaims) {
+          hasClaim(requireClaims, userClaims) && filteredRoutes.push(route);
+        } else {
+          if (typeof requireParentClaims === "object") {
+            hasClaim(requireParentClaims, userClaims) &&
+              filteredRoutes.push(route);
+          } else {
+            filteredRoutes.push(route);
+          }
+        }
+      }
+      return filteredRoutes;
+    },
+    [] as RouteConfig[]
+  );
+};
+
 export const routerBeforeEach: NavigationGuard = async (to, from, next) => {
   // reverse routes so nested routes can take control
-  const requireAuthRoute = [...to.matched]
+  const restrictedRoute = [...to.matched]
     .reverse()
     .find(({ meta }) => meta.auth);
 
-  if (!requireAuthRoute) return next();
+  if (!restrictedRoute) return next();
 
   const currentUser = firebase.auth().currentUser;
   const {
     meta: {
       auth: { requireAuth, guestOnly, requireClaims }
     }
-  } = requireAuthRoute;
+  } = restrictedRoute;
 
   if (requireAuth && !currentUser)
-    next({
+    return next({
       path: "/login",
       query: { redirect: to.fullPath }
     });
-  else if (guestOnly && currentUser) next("/");
-  else if (requireClaims) {
+
+  if (guestOnly && currentUser) return next("/");
+
+  if (requireClaims) {
     const userClaims = currentUser
       ? (await currentUser.getIdTokenResult()).claims
       : null;
 
-    if (!userClaims) next("/restricted");
-    else {
-      const hasRequiredClaim = Object.entries(requireClaims).some(
-        ([claimName, claimValue]) => {
-          return userClaims[claimName] === claimValue;
-        }
-      );
+    return !userClaims || !hasClaim(requireClaims, userClaims)
+      ? next("/")
```

Comment:
Now I'm thinking, maybe we can just abort the transition? As [documentation](https://router.vuejs.org/guide/advanced/navigation-guards.html#global-guards) says, we can call `next(false)` or `next(error)`. The only question is: what happens if the route is accessed by a direct link?

Reply:
Using next(false) appears to show a blank page when accessed by direct link.

Reply:
Let it redirect to the home page, as it is now.